### ISSUE TITLE
[BUG] Lack of error reporting when unevaluated graph becomes too big

### ISSUE THREAD

**Issue by aristidb** [ml-explore/mlx#1712]
**Describe the bug**
When failing to evaluate causes the graph to become excessively big, I would expect to get a clear error message.

Instead, it:

1. Reports no error
2. CPU & GPU usage goes to 0%
3. It just hangs indefinitely (at least 2 hours) in Metal Malloc (and no this can't be explained by evaluating the graph lazily taking a long time, as it takes only 143 seconds when running with the correct evals)

**To Reproduce**

Include code snippet
```python
import mlx.core as mx
import mlx.nn as nn
import time

Xtr = mx.zeros((182625, 3), dtype=mx.int32)
Ytr = mx.zeros((182625,), dtype=mx.int32)

mx.random.seed(7674764)
C = mx.random.normal((27, 10))
W1 = mx.random.normal((30, 200))
b1 = mx.random.normal((200,))
W2 = mx.random.normal((200, 27))
b2 = mx.random.normal((27,))
parameters = [C, W1, b1, W2, b2]

def loss_fn(X: mx.array, Y: mx.array, C: mx.array, W1: mx.array, b1: mx.array, W2: mx.array, b2: mx.array):
    h = mx.tanh(C[X].flatten(1) @ W1 + b1)
    logits = h @ W2 + b2
    return nn.losses.cross_entropy(logits, Y, reduction='mean')
loss_value_and_grad = mx.value_and_grad(loss_fn, range(2, 7))
loss_value_and_grad = mx.compile(loss_value_and_grad)

start = time.time()
for i in range(200000):
    time.sleep(0.001)
    ix = mx.random.randint(0, Xtr.shape[0], (320,))  # minibatch
    loss, grad = loss_value_and_grad(Xtr[ix], Ytr[ix], C, W1, b1, W2, b2)
    lr = 0.2 if i < 100000 else 0.02
    for pi, p in enumerate(parameters):
        p -= lr * grad[pi]
```

**Expected behavior**
Clear error message, or finish in reasonable amount of time.

**Desktop (please complete the following information):**
 - OS Version: macOS 15.2
 - Version: 0.21.1

**Additional context**
I couldn't reopen #1711 after it was closed, so I'm just making a new issue.

**Comment by awni** [comment]
Apple devices have a built-in resource limit (~500k) which is configurable by setting the sysctl `iogpu.rsrc_limit`. For some reason the metal system allocator hangs if the resource limit is exceeded 🤷 .

We've run into this a few times before and there is no useful error messaging. #1718 should now throw on your test case rather than hanging.

**Comment by aristidb** [comment]
Thanks, I agree that #1718 should solve my main concern once it is merged

**Comment by tianenchong** [comment]
Would it be a good feature to include current `mlx::core::metal::MetalAllocator::num_resources_` monitoring?
